### PR TITLE
[Merged by Bors] - feat: Mixins for monotonicity of scalar multiplication

### PR DESCRIPTION
**Pull request by YaelDillies**
This PR introduces eight typeclasses for monotonicity of left/right scalar multiplication by nonnegative elements:
* `PosSMulMono`: If `a ≥ 0`, then `b₁ ≤ b₂` implies `a • b₁ ≤ a • b₂`.
* `PosSMulStrictMono`: If `a > 0`, then `b₁ < b₂` implies `a • b₁ < a • b₂`.
* `PosSMulReflectLT`: If `a ≥ 0`, then `a • b₁ < a • b₂` implies `b₁ < b₂`.
* `PosSMulReflectLE`: If `a > 0`, then `a • b₁ ≤ a • b₂` implies `b₁ ≤ b₂`.
* `SMulPosMono`: If `b ≥ 0`, then `a₁ ≤ a₂` implies `a₁ • b ≤ a₂ • b`.
* `SMulPosStrictMono`: If `b > 0`, then `a₁ < a₂` implies `a₁ • b < a₂ • b`.
* `SMulPosReflectLT`: If `b ≥ 0`, then `a₁ • b < a₂ • b` implies `a₁ < a₂`.
* `SMulPosReflectLE`: If `b > 0`, then `a₁ • b ≤ a₂ • b` implies `a₁ ≤ a₂`.

The design is heavily inspired to the corresponding one for multiplication (see `Algebra.Order.Ring.Lemmas`). Note however the following differences:
* The new typeclasses are custom typeclasses instead of abbreviations for the correct `CovariantClass`/`ContravariantClass` invokations. This has the following benefits:
  * They get displayed as classes in the docs. In particular, one can see their list of instances, instead of their instances being invariably dumped to the `CovariantClass`/`ContravariantClass` list.
  * They don't pollute other typeclass searches. Having many abbreviations of the same typeclass for different purposes always felt like a performance issue to me (more instances with the same key, for no added benefit), and indeed a previous version of this PR hit timeouts due to the higher number of `CovariantClass`/`ContravariantClass` instances.
  * `SMulPosReflectLT`/`SMulPosReflectLE` did not fit in the framework since they relate `≤` on two different types. So I would have had to generalise `CovariantClass`/`ContravariantClass` to three types and two relations.
  * Very minor, but the constructors let you work with `a : α`, `h : 0 ≤ a` instead of `a : {a : α // 0 ≤ a}`. This actually makes some instances surprisingly cleaner to prove.
  * The `CovariantClass`/`ContravariantClass` framework was only used to automate very simple logic anyway. It was easily copied over.
* We replace undocumented lemmas stating the equivalence of the four typeclasses mentioning nonnegativity with their positivity version by motivated constructors.
* We abandon series of lemmas of dubious utility. Those were already marked as such in `Algebra.Order.Ring.Lemmas` (by myself).
* Some lemmas about commutativity of multiplication don't make sense for scalar multiplication.

This PR links the new typeclasses to `OrderedSMul` and makes all old lemmas in `Algebra.Order.SMul` one-liners in terms of the new lemmas (except when they have the same name, in which case the lemma is simply moved) but doesn't delete the old ones to reduce churn. What remains to be done afterwards is thus:
* finish the transition by deleting the duplicate lemmas from `Algebra.Order.SMul`
* rearrange the non-duplicate lemmas into new files
* generalise (most of) the lemmas from `Algebra.Order.Module` to `Algebra.Order.Module.Defs`
* rethink `OrderedSMul`

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Order/Module/Defs.lean [hunk 1]
```diff
@@ -0,0 +1,826 @@
...
+This file acts as a substitute for `Mathlib.Algebra.Order.SMul`. We now need to
+* finish the transition by deleting the duplicate lemmas
+* rearrange the non-duplicate lemmas into new files
+* generalise (most of) the lemmas from `Mathlib.Algebra.Order.Module` to here
+* rethink `OrderedSMul`
+-/
+
+variable (α β : Type*)
+
+section Abbreviations
```
Comment (eric-wieser) [inline Mathlib/Algebra/Order/Module/Defs.lean:87]: There are no abbreviations here.

Can you add an implementation note that compares why we use dedicated classes here, but abbreviations for the `Mul` versions?

Comment (YaelDillies) [reply Mathlib/Algebra/Order/Module/Defs.lean:87]: There's not really any reason to use `abbrev` for the `Mul` versions, but "Let's keep the diff small".

Comment (YaelDillies) [reply Mathlib/Algebra/Order/Module/Defs.lean:87]: Copied over the blurb from my PR description

**Comment by eric-wieser** [comment]
> * `...MonoRev` is replaced by `...ReflectLE` in the hope of being more understandable.

Can you spin out a PR to make this change to the `Mul` typeclasses first?

**Comment by YaelDillies** [comment]
#8890

**Comment by jcommelin** [comment]
!bench

**Comment by jcommelin** [comment]
Hmm, that `!bench` isn't going to measure anything, because this new API isn't really used yet.

bors merge

**Pull request closed without merging**